### ISSUE TITLE
Error when using init='random' in torch 1.7.0

### ISSUE THREAD

**Issue by RichieHakim** [tensorly/tensorly#200]
#### Describe the bug

Error when using init='random' with pytorch backend and RTX 3090 GPU
This occurred after upgrading from torch 1.6.0 to 1.7.0
The below code works when switching to init='svd', or when using 'cpu' device

The input tensors I am using for CP decomposition are far too big to initialize with 'svd', so 'random' is necessary

#### Versions
python 3.8.5 (default, Sep  4 2020, 07:30:14) 
torch 1.7.0
CUDA 11.1

using pytorch backend

Linux-5.4.0-54-generic-x86_64-with-glibc2.10
[GCC 7.3.0]
NumPy 1.19.2
SciPy 1.5.2
TensorLy 0.5.0

#### Steps or Code to Reproduce

import numpy as np
import torch
import torch.cuda
import tensorly as tl
import tensorly.decomposition
import tensorly.random

cuda_device_number = torch.cuda.current_device()
device = f'cuda:{cuda_device_number}'

data = np.random.rand(100,90,80,70)

tl.set_backend('pytorch')
data_tensor = tl.tensor(data, dtype=tl.float32, device='cuda:0', requires_grad=False)
weights, factors = tl.decomposition.non_negative_parafac(data_tensor, init='random', tol=1e-05, n_iter_max=100, rank=3, verbose=True)

#### Actual result

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-79af35d8d396> in <module>
     17 tl.set_backend('pytorch')
     18 data_tensor = tl.tensor(data, dtype=tl.float32, device='cuda:0', requires_grad=False)
---> 19 weights, factors = tl.decomposition.non_negative_parafac(data_tensor, tol=1e-05, init='random', n_iter_max=100, rank=3, verbose=True)
     20 

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/decomposition/_cp.py in non_negative_parafac(tensor, rank, n_iter_max, init, svd, tol, random_state, verbose, normalize_factors, return_errors, mask, orthogonalise, cvg_criterion, fixed_modes)
    512                 tensor = tensor*mask + tl.cp_to_tensor((None, factors), mask=1-mask)
    513 
--> 514             mttkrp = unfolding_dot_khatri_rao(tensor, (None, factors), mode)
    515 
    516             numerator = tl.clip(mttkrp, a_min=epsilon, a_max=None)

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/cp_tensor.py in unfolding_dot_khatri_rao(tensor, cp_tensor, mode)
    533     weights, factors = cp_tensor
    534     for r in range(rank):
--> 535         component = multi_mode_dot(tensor, [f[:, r] for f in factors], skip=mode)
    536         mttkrp_parts.append(component)
    537 

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/tenalg/__init__.py in dynamically_dispatched_fun(*args, **kwargs)
     78         current_backend = _BACKENDS[_LOCAL_STATE.tenalg_backend]
     79         if hasattr(current_backend, name):
---> 80             fun = getattr(current_backend, name)(*args, **kwargs)
     81         else:
     82             warnings.warn(f'tenalg: defaulting to core tenalg backend, {name}'

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/tenalg/core_tenalg/n_mode_product.py in multi_mode_dot(tensor, matrix_or_vec_list, modes, skip, transpose)
    121             res = mode_dot(res, T.conj(T.transpose(matrix_or_vec)), mode - decrement)
    122         else:
--> 123             res = mode_dot(res, matrix_or_vec, mode - decrement)
    124 
    125         if T.ndim(matrix_or_vec) == 1:

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/tenalg/core_tenalg/n_mode_product.py in mode_dot(tensor, matrix_or_vector, mode)
     62                              'Provided array of dimension {} not in [1, 2].'.format(T.ndim(matrix_or_vector)))
     63 
---> 64         res = T.dot(matrix_or_vector, unfold(tensor, mode))
     65 
     66         if vec: # We contracted with a vector, leading to a vector

~/miniconda3/envs/torch_env/lib/python3.8/site-packages/tensorly/backend/__init__.py in inner(*args, **kwargs)
    159 
    160     def inner(*args, **kwargs):
--> 161         return _get_backend_method(name)(*args, **kwargs)
    162 
    163     # We don't use `functools.wraps` here because some of the dispatched

RuntimeError: Tensor for 'out' is on CPU, Tensor for argument #1 'self' is on CPU, but expected them to be on GPU (while checking arguments for addmm)

**Comment by JeanKossaifi** [comment]
Thanks for the detailed issue @RichieHakim! 

The issue was due to the random initialization not using the same device as the input tensor -- this has been fixed in the master branch by bd34920ec1d8a54e97fd7f6bafb500f2e520a6ce: if you use the dev (Github) version this issue should be solved.

As a side note, you can wrap the Traceback in backticks ``` to get a more legible rendering on Github:
```ipython
RuntimeError Traceback (most recent call last)
in
17 tl.set_backend('pytorch')
18 data_tensor = tl.tensor(data, dtype=tl.float32, device='cuda:0', requires_grad=False)
---> 19 weights, factors = tl.decomposition.non_negative_parafac(data_tensor, tol=1e-05, init='random', n_iter_max=100, rank=3, verbose=True)
```

Closing the issue, feel free to reopen if you still observe the issue.

**Comment by RichieHakim** [comment]
Works brilliantly, thank you!
